File: matlab/+core/adaptivity.m
function result = adaptivity(problem, initial_mesh, tol_time, tol_PA, tol_delta, max_iters)
%SOLVE_OPTIMAL_CONTROL Adaptive solver for optimal control problems.
%   result = solve_optimal_control(problem, initial_mesh, tol_time, tol_PA, tol_delta, max_iters)
%   runs the outer adaptivity loop described in Algorithm 3.1 of the
%   DeepResearch plan.  It iteratively solves the TPBVP, computes
%   a posteriori error indicators for time discretisation, PA surrogate
%   error and smoothing error, and refines the discretisation until
%   tolerances are met.  The returned struct 'result' contains the
%   final trajectories X,P, control sequence u, time mesh and logs.

% Set default tolerances if missing
if nargin < 3 || isempty(tol_time);  tol_time = 1e-3; end
if nargin < 4 || isempty(tol_PA);    tol_PA   = 1e-3; end
if nargin < 5 || isempty(tol_delta); tol_delta= 1e-3; end
if nargin < 6 || isempty(max_iters); max_iters = 20; end

% Initialise mesh, bundle and smoothing parameter
t_nodes = initial_mesh(:)';
if t_nodes(1) ~= 0
    error('Initial mesh must start at t=0');
end
% If final time unspecified in problem, use last node of initial mesh
if isempty(problem.T)
    problem.T = t_nodes(end);
elseif abs(problem.T - t_nodes(end)) > 1e-12
    % adjust mesh to end at specified T
    t_nodes(end) = problem.T;
end
% Initialise PA bundle with one control candidate: zero or lower bound
m = length(problem.getControlBounds());
[amin, amax] = problem.getControlBounds();
if isempty(amin)
    init_ctrl = zeros(0,1);
else
    init_ctrl = (amin + amax)/2;
end
bundle = core.PABundle(init_ctrl);
% Initial smoothing parameter delta
delta = 0.1;
% Initial guess for state and costate trajectories
n = length(problem.x0);
N = length(t_nodes) - 1;
Xinit = repmat(problem.x0(:), 1, N+1);
Pinit = zeros(n, N+1);

log = struct();
log.iter = [];
log.N  = [];
log.M  = [];
log.delta = [];
log.eta_time = [];
log.eta_PA   = [];
log.eta_delta= [];

for k = 1:max_iters
    % Solve TPBVP on current discretisation
    opts = struct('tol', 1e-8, 'maxIter', 50, 'alpha', 1e-4, 'beta', 0.5, 'lam_min', 1e-8);
    [X, P, info] = core.newton().solve_tpbvp(problem, bundle, delta, t_nodes, Xinit, Pinit, opts);
    % Compute a posteriori error indicators
    [eta_time, local_time_err] = estimate_time_error(problem, bundle, delta, t_nodes, X, P);
    eta_PA   = estimate_PA_error(problem, bundle, t_nodes, X, P);
    eta_delta= estimate_smoothing_error(problem, bundle, delta, t_nodes, X, P);
    % Log iteration info
    log.iter(end+1)     = k;
    log.N(end+1)        = length(t_nodes)-1;
    log.M(end+1)        = bundle.num_planes();
    log.delta(end+1)    = delta;
    log.eta_time(end+1) = eta_time;
    log.eta_PA(end+1)   = eta_PA;
    log.eta_delta(end+1)= eta_delta;
    % Check stopping criteria
    if (eta_time <= tol_time) && (eta_PA <= tol_PA) && (eta_delta <= tol_delta)
        break;
    end
    % Determine refinement: priority: time, PA, delta
    if eta_time > tol_time
        % Refine mesh: split intervals where local error > tol_time
        old_nodes = t_nodes;
        t_nodes = refine_time_mesh(t_nodes, local_time_err, tol_time);
        % Interpolate X and P from old mesh to new mesh by simple linear interpolation
        [Xinit, Pinit] = interpolate_to_new_mesh(t_nodes, X, P, old_nodes);
    elseif eta_PA > tol_PA
        % Add plane: find worst gap and add control
        [t_idx, a_new] = find_worst_plane(problem, bundle, t_nodes, X, P);
        bundle = bundle.add_control(a_new);
        % Reuse current X,P as initial guess
        Xinit = X;
        Pinit = P;
    elseif eta_delta > tol_delta
        % Reduce smoothing
        delta = delta / 2;
        % Reuse current X,P
        Xinit = X;
        Pinit = P;
    else
        % Should not reach; but ensure progress
        delta = delta / 2;
        Xinit = X;
        Pinit = P;
    end
end

% Extract control trajectory
N = length(t_nodes)-1;
m = size(bundle.controls,1);
u = zeros(m, N+1);
for i = 1:N+1
    p_i = P(:,i);
    x_i = X(:,i);
    t_i = t_nodes(i);
    [~, a_star] = core.hamiltonian.compute_H(problem, p_i, x_i, t_i, bundle, false);
    u(:,i) = a_star;
end

% Package result
result = struct();
result.t   = t_nodes;
result.X   = X;
result.P   = P;
result.u   = u;
result.bundle = bundle;
result.delta  = delta;
result.log    = log;
result.iterations = k;
end

function [eta_time, local_err] = estimate_time_error(problem, bundle, delta, t_nodes, X, P)
%ESTIMATE_TIME_ERROR Estimate the time discretisation error.
%   Returns a global indicator eta_time and per‑interval error local_err.
N = length(t_nodes) - 1;
n = size(X,1);
local_err = zeros(1, N);
% Use difference of gradient of H_delta across interval as proxy for
% curvature; larger changes suggest insufficient time resolution.
for i = 1:N
    ti   = t_nodes(i);
    tip1 = t_nodes(i+1);
    xi   = X(:,i);
    xip1 = X(:,i+1);
    pi   = P(:,i);
    pip1 = P(:,i+1);
    [~, gradp_i, ~] = core.smoothing.eval_H_smooth(bundle, problem, pi, xi, ti, delta);
    [~, gradp_ip1, ~] = core.smoothing.eval_H_smooth(bundle, problem, pip1, xip1, tip1, delta);
    dt = tip1 - ti;
    % Error estimate: change in gradient multiplied by dt
    local_err(i) = norm(gradp_ip1 - gradp_i, 2) * dt;
end
% Global error indicator: maximum local error
eta_time = max(local_err);
end

function eta_PA = estimate_PA_error(problem, bundle, t_nodes, X, P)
%ESTIMATE_PA_ERROR Estimate the model (bundle) error.
%   eta_PA = ∫ (\bar{H} - H) dt approximated by trapezoid rule.
N = length(t_nodes) - 1;
vals = zeros(1, N+1);
for i = 1:N+1
    p_i = P(:,i);
    x_i = X(:,i);
    t_i = t_nodes(i);
    % Surrogate Hbar
    [Hbar, ~] = bundle.eval(problem, p_i, x_i, t_i);
    % True H_K (state constraints) – restrict to tangent cone
    [Htrue, ~] = core.hamiltonian.compute_H(problem, p_i, x_i, t_i, bundle, true);
    vals(i) = Hbar - Htrue;
end
% Integrate by trapezoidal rule
eta_PA = sum(0.5 * (vals(1:end-1) + vals(2:end)) .* diff(t_nodes));
end

function eta_delta = estimate_smoothing_error(problem, bundle, delta, t_nodes, X, P)
%ESTIMATE_SMOOTHING_ERROR Estimate the smoothing bias error.
%   eta_delta = ∫ (H_delta - \bar{H}) dt approximated by trapezoid rule.
N = length(t_nodes) - 1;
vals = zeros(1, N+1);
for i = 1:N+1
    p_i = P(:,i);
    x_i = X(:,i);
    t_i = t_nodes(i);
    [Hdelta, ~, ~] = core.smoothing.eval_H_smooth(bundle, problem, p_i, x_i, t_i, delta);
    [Hbar, ~] = bundle.eval(problem, p_i, x_i, t_i);
    vals(i) = Hdelta - Hbar;
end
eta_delta = sum(0.5 * (vals(1:end-1) + vals(2:end)) .* diff(t_nodes));
end

function new_mesh = refine_time_mesh(t_nodes, local_err, tol)
%REFINE_TIME_MESH Refine time mesh based on local error indicators.
%   Splits intervals where the local error exceeds tol by inserting the
%   midpoint.  Returns the new sorted mesh.
N = length(t_nodes) - 1;
new_mesh = t_nodes;
for i = 1:N
    if local_err(i) > tol
        mid = (t_nodes(i) + t_nodes(i+1)) / 2;
        new_mesh = [new_mesh, mid];
    end
end
new_mesh = sort(new_mesh);
% Remove duplicates (within numerical tolerance)
tol_time = 1e-12;
new_mesh = unique(round(new_mesh/ tol_time)*tol_time);
end

function [Xnew, Pnew] = interpolate_to_new_mesh(t_new, X, P, t_old)
%INTERPOLATE_TO_NEW_MESH Interpolate X and P to a refined mesh.
%   [Xnew, Pnew] = interpolate_to_new_mesh(t_new, X, P, t_old) returns
%   the state and costate trajectories Xnew and Pnew on the new mesh
%   `t_new`, given the trajectories X and P on the old mesh `t_old`.
%   Linear interpolation is used for each state and costate component.

    n = size(X,1);
    Xnew = zeros(n, length(t_new));
    Pnew = zeros(n, length(t_new));
    for j = 1:n
        Xnew(j,:) = interp1(t_old, X(j,:), t_new, 'linear');
        Pnew(j,:) = interp1(t_old, P(j,:), t_new, 'linear');
    end
    % Ensure exact x0 at t=0 if present in old mesh
    if abs(t_new(1) - t_old(1)) < 1e-12
        Xnew(:,1) = X(:,1);
    end
end

function [t_idx, a_new] = find_worst_plane(problem, bundle, t_nodes, X, P)
%FIND_WORST_PLANE Identify time index and control that maximally reduces gap.
%   Evaluates Hbar - Htrue at each node and returns the index of the
%   largest gap along with the true minimizer a_new to add as plane.
N = length(t_nodes) - 1;
gaps = zeros(1, N+1);
a_cands = zeros(length(bundle.controls(:,1)), N+1);
for i = 1:N+1
    p_i = P(:,i);
    x_i = X(:,i);
    t_i = t_nodes(i);
    [Hbar, ~] = bundle.eval(problem, p_i, x_i, t_i);
    [Htrue, a_star] = core.hamiltonian.compute_H(problem, p_i, x_i, t_i, bundle, true);
    gaps(i) = Hbar - Htrue;
    a_cands(:,i) = a_star;
end
[~, t_idx] = max(gaps);
a_new = a_cands(:,t_idx);
end

File: python/src/core/adaptivity.py
import numpy as np

from .pa_bundle import PABundle
from .smoothing import eval_H_smooth
from .hamiltonian import compute_H
from .newton import solve_tpbvp

def solve_optimal_control(
    problem,
    initial_mesh: np.ndarray,
    tol_time: float = 1e-3,
    tol_PA: float = 1e-3,
    tol_delta: float = 1e-3,
    max_iters: int = 10,
    delta0: float = 0.1,
):
    """
    Solve an optimal control problem adaptively by refining time mesh,
    adding new control planes, and reducing the smoothing parameter.

    Parameters
    ----------
    problem : OCPProblem
        Problem definition.
    initial_mesh : np.ndarray
        Initial time grid (including 0 and T).  Should be sorted.
    tol_time : float
        Tolerance for the time discretisation error indicator.
    tol_PA : float
        Tolerance for the PA surrogate error indicator.
    tol_delta : float
        Tolerance for the smoothing error indicator.
    max_iters : int
        Maximum number of outer adaptivity iterations.
    delta0 : float
        Initial smoothing parameter.

    Returns
    -------
    dict
        Dictionary with solution, mesh, bundle, delta, and log information.
    """
    # copy mesh
    t_nodes = np.asarray(initial_mesh, dtype=float).copy()
    # initialize PA bundle with zero control if dimension known, otherwise empty
    bundle = PABundle()
    # try to add zero control (or mean of bounds) if possible
    bounds = problem.control_bounds_tuple()
    m = problem.m
    if m is None and bounds is not None:
        m = bounds[0].size
    if m is not None:
        if bounds is not None:
            u0 = 0.5 * (bounds[0] + bounds[1])
        else:
            u0 = np.zeros(m)
        bundle.add_control(u0)
    delta = delta0
    log = []
    # initial guesses for X and P: None (will be set in Newton)
    X_guess = None
    P_guess = None
    for k in range(max_iters):
        # solve TPBVP on current mesh with current bundle and delta
        X, P, info = solve_tpbvp(problem, t_nodes, bundle, delta, X_guess, P_guess)
        # compute error indicators
        # time discretisation error
        N = len(t_nodes) - 1
        n = problem.n
        eta_time_local = np.zeros(N)
        grad_p_list = []
        grad_x_list = []
        # compute gradients at nodes for time error and smoothing error
        for i in range(N + 1):
            # grad_p, grad_x at each node
            # use smoothing evaluation
            _, grad_p_i, grad_x_i = eval_H_smooth(problem, bundle, P[i], X[i], t_nodes[i], delta)
            grad_p_list.append(grad_p_i)
            grad_x_list.append(grad_x_i)
        # compute local error indicator as difference of gradient across interval
        for i in range(N):
            dt = t_nodes[i + 1] - t_nodes[i]
            gp0 = grad_p_list[i]
            gp1 = grad_p_list[i + 1]
            gx0 = grad_x_list[i]
            gx1 = grad_x_list[i + 1]
            # measure change
            diff_gp = gp1 - gp0
            diff_gx = gx1 - gx0
            eta_time_local[i] = dt * (np.linalg.norm(diff_gp) + np.linalg.norm(diff_gx))
        eta_time = np.max(eta_time_local) if N > 0 else 0.0
        # PA error: integrate (Hbar - H)
        eta_PA = 0.0
        for i in range(N):
            # at node i and i+1, compute gap
            Hbar_i, _ = bundle.evaluate(problem, P[i], X[i], t_nodes[i])
            Hbar_ip1, _ = bundle.evaluate(problem, P[i + 1], X[i + 1], t_nodes[i + 1])
            # compute true H (restricted) at i and i+1
            H_i, _ = compute_H(problem, P[i], X[i], t_nodes[i], bundle.controls, restricted=True)
            H_ip1, _ = compute_H(problem, P[i + 1], X[i + 1], t_nodes[i + 1], bundle.controls, restricted=True)
            gap_i = Hbar_i - H_i
            gap_ip1 = Hbar_ip1 - H_ip1
            dt = t_nodes[i + 1] - t_nodes[i]
            eta_PA += 0.5 * (gap_i + gap_ip1) * dt
        # smoothing error: integrate (H_delta - Hbar)
        eta_delta = 0.0
        for i in range(N):
            Hdelta_i, _, _ = eval_H_smooth(problem, bundle, P[i], X[i], t_nodes[i], delta)
            Hdelta_ip1, _, _ = eval_H_smooth(problem, bundle, P[i + 1], X[i + 1], t_nodes[i + 1], delta)
            Hbar_i, _ = bundle.evaluate(problem, P[i], X[i], t_nodes[i])
            Hbar_ip1, _ = bundle.evaluate(problem, P[i + 1], X[i + 1], t_nodes[i + 1])
            diff_i = Hdelta_i - Hbar_i
            diff_ip1 = Hdelta_ip1 - Hbar_ip1
            dt = t_nodes[i + 1] - t_nodes[i]
            eta_delta += 0.5 * (diff_i + diff_ip1) * dt
        log.append({
            'iteration': k,
            'N': N,
            'M': bundle.num_planes(),
            'delta': delta,
            'eta_time': eta_time,
            'eta_PA': eta_PA,
            'eta_delta': eta_delta,
            'newton_iter': info['iterations'],
            'newton_residual': info['residual_norm'],
        })
        # check convergence
        if (eta_time <= tol_time) and (eta_PA <= tol_PA) and (eta_delta <= tol_delta):
            break
        # priority: refine time first, then PA planes, then reduce delta
        if eta_time > tol_time:
            # refine time mesh: subdivide intervals with high local error
            new_nodes = [t_nodes[0]]
            X_new = [X[0]]
            P_new = [P[0]]
            for i in range(N):
                dt = t_nodes[i + 1] - t_nodes[i]
                # compute midpoint and error indicator
                err = eta_time_local[i]
                if err > tol_time:
                    # insert midpoint
                    t_mid = 0.5 * (t_nodes[i] + t_nodes[i + 1])
                    # linear interpolate X and P
                    alpha = (t_mid - t_nodes[i]) / dt
                    x_mid = (1 - alpha) * X[i] + alpha * X[i + 1]
                    p_mid = (1 - alpha) * P[i] + alpha * P[i + 1]
                    new_nodes.extend([t_mid])
                    X_new.extend([x_mid])
                    P_new.extend([p_mid])
                new_nodes.append(t_nodes[i + 1])
                X_new.append(X[i + 1])
                P_new.append(P[i + 1])
            t_nodes = np.array(new_nodes, dtype=float)
            X_guess = np.array(X_new)
            P_guess = np.array(P_new)
            continue
        if eta_PA > tol_PA:
            # add new plane: find worst gap index
            max_gap = -np.inf
            max_idx = 0
            for i in range(N + 1):
                Hbar_i, _ = bundle.evaluate(problem, P[i], X[i], t_nodes[i])
                H_i, u_star = compute_H(problem, P[i], X[i], t_nodes[i], bundle.controls, restricted=True)
                gap = Hbar_i - H_i
                if gap > max_gap:
                    max_gap = gap
                    max_idx = i
                    best_u = u_star
            # add best_u to bundle
            if best_u is not None:
                bundle.add_control(best_u)
            X_guess = X
            P_guess = P
            continue
        # else reduce delta
        if eta_delta > tol_delta:
            delta = delta * 0.5
            # do not change mesh or bundle
            X_guess = X
            P_guess = P
            continue
    # return final solution and log
    return {
        't_nodes': t_nodes,
        'X': X,
        'P': P,
        'bundle': bundle,
        'delta': delta,
        'log': log
    }